### ISSUE TITLE
TODO list for integration of app-model

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#4860]
## 🧰 Task
list of todos after #4784 

- [ ] continue replacing procedural QtMenus with app-model registered ones (note PR #4826 provides the base changes and tests required for the menu migration to app-model and all other PRs are based off of this original PR. It should thus be merged first)
    - [x] view menu: #4826 
    - [x] file menu: #4865
    - [x] help menu: #4922 
    - [x] window menu: https://github.com/napari/napari/pull/6905
    - [x] plugins menu:  #4991
    - [x] Debug: #6915
- [x] allow registration of app keybindings that aren't attached to a specific python object (like viewer or layers).  I think this could be done gradually.  to start, we could by put a object that implements the  `KeymapProvider` interface at the beginning or end of the `ChainMap` used to lookup keybindings.  In the longer run, the [app-model keybindings registry](https://app-model.readthedocs.io/en/latest/registries/) would be the source of keybindings, and `key_bindings.KeymapHandler` would be retired.
- [ ] extract keypress control from the vispy canvas (https://github.com/napari/napari/issues/3087) 
- [ ] add documentation to `napari.org/plugins` that explains how to add menu items, including a list of "contributable" menu ids, (similar to [this](https://code.visualstudio.com/api/references/contribution-points#contributes.menus))
- [ ] add documentation to `napari.org/plugins` explaining exactly what names they can use in their `when` context clauses in their manifests (like [this page for vscode](https://code.visualstudio.com/api/references/when-clause-contexts)). Related issue: https://github.com/napari/docs/issues/146, also for reference Talley has mentioned wanting to (auto)generate docs using the descriptions in the ContextKeys: https://github.com/napari/napari/pull/4499#issuecomment-1129426998
- [ ] add documentation explaining how to use `superqt.fonticons` to declare icons (and figure out the best way for those icons to change color with the theme)
- [ ] decide on and add documentation to explain exactly what types may be used as type hints for dependency injection in a plugin command function.  (these are the names that are declared in the the `app.injection_store.namespace`)
- [ ] replace action manager with commands registry, determine if any deprecation is needed 
- [ ] consider putting the notification system in the app-model (perhaps upstream)
- [ ] create 3 tiers of keybinding scope: default (napari), plugins, and user keybindings (in order of increasing priority)
- [ ] rename `_qt.get_app` to `get_qapp`... however that's publicly exported, and that needs to be considered
- [ ] consider using Actions for layer mode buttons

- [ ] move to expressing keybindings using [app-models keycodes](https://app-model.readthedocs.io/en/latest/keybindings/) instead of vispy's  (the app-model tests and demos have a lot of examples of how to express keypresses using the `KeyMod` and `KeyCode` enums)
- [ ] #4919
- [ ] https://github.com/napari/napari/issues/4979

**Comment by liu-ziyang** [comment]
@tlambert03 Thanks for putting this together! do you mind if I convert this issue to a project board so we can track and collaborate?

**Comment by Czaki** [comment]
> * [ ]  rename `_qt.get_app` to `get_qapp`... however that's publicly exported, and that needs to be considered

the `napari.qt` could still export `get_app` with a deprecation warning.

**Comment by lucyleeow** [comment]
Small addition:

Update `docs/guides/contexts_expressions` to explain their use in context of app-model.

(I am assuming that after full transition to app-model we won't be using the old system at all anymore?)

**Comment by lucyleeow** [comment]
Documenting some planning/tasks related to app-model integration.

Tasks post #4991 :

* Fix for loop in npe1 samples menu  `_rebuild_npe1_samples_menu`
* Use `partial` in `_rebuild_npe1_samples_menu`
* https://github.com/napari/napari/issues/6516 - in first instance remove use of enums for commandID and menuIDs (which removes 2 files)
* https://github.com/napari/napari/issues/6585
* Remove lambdas in `Window._add_menus` (see: https://github.com/napari/napari/pull/4991#discussion_r1429848050 for details)

Other related tasks, where I am not sure of priority/order to be worked on:

* https://github.com/pyapp-kit/app-model/issues/52
* https://github.com/pyapp-kit/app-model/issues/176
* https://github.com/napari/napari/pull/4977
* https://github.com/napari/napari/issues/6326 (this may be fixed with menu contributions / nap 6 implementation)
* Split https://github.com/napari/napari/pull/6204 (note action manager deprecation occurs here)

Post npe1 deprecation tasks (only listing items relevant/related to this body of work):

*  deprecate npe1 when adding widgets in `_run` and deprecate `_add_plugin_function_widget` and `add_plugin_dock_widget` (no function/dock widget distinction in npe2
* remove npe1 widget tests

Lower priority (?) related tasks:

* https://github.com/napari/napari/issues/6524
* https://github.com/napari/napari/issues/6232

cc @Czaki @DragaDoncila in case I missed something.

**Comment by DragaDoncila** [comment]
@lucyleeow I think #6643 should also be listed somewhere in here, probably in the lower priority tasks, since we have a workaround for now

**Comment by lucyleeow** [comment]
Done and added some more!
